[libc] Updated header_generation.rst (#99712)
llvm/llvm-project: libc/docs/dev/header_generation.rst

> .. role:: raw-html(raw)

Reviewer: is this needed?

Author: It is required to force line break in step 5 of Common Errors. I can't use line block syntax because there is a vertical line in the text, and escape character did not work there.

    :format: html

Generating Public and Internal headers
======================================

> This is a new implementation of the previous libc header generator. The old

Reviewer: we should talk about old headergen in the past tense (e.g. "it was")

header generator (libc-hdrgen aka “headergen”) is based on tablegen, which
creates an awkward dependency on the rest of LLVM for our build system. By
creating a new standalone headergen we can eliminate these dependencies for
easier cross compatibility.

There are 3 main components of the new Headergen. The first component are the
yaml files that contain all the function header information and are separated by
header specification and standard. The second component are the classes that are
created for each component of the function header: macros, enumerations, types,
function, arguments, and objects. The third component is the python script that
uses the class representation to deserialize yaml files into its specific
components and then reserializes the components into the function header. The
python script also combines the generated header content with header definitions
and extra macro and type inclusions from the .h.def file.

Instructions

> -----------------------------

Reviewer: nit: this line should be the same length as the one above it.

Required Versions:

> - Python Version: 3.11.8 [subject to be lower]

Reviewer: we can just leave the supported python version as 3.11.8. It will probably work with earlier versions but I don't think we need to guarantee that.

- CMake Version: 3.20.0

1. Make sure to have `LLVM <https://llvm.org/docs/GettingStarted.html>`_ on your
   system.
2. Make sure you have created a build directory within your LLVM directory.
3. When generating the necessary files by your build make sure to include:
   ``-DLLVM_LIBC_FULL_BUILD=ON`` within the CMake command since new headergen
   only runs on full-build.
4. Once the build is complete, enter in the command line within the build
   directory ``ninja check-newhdrgen`` to ensure that the integration tests are
   passing.
5. Then enter in the command line ``ninja libc`` to generate headers. Headers
   will be in “build/projects/libc/include” or “build/libc/include” in a runtime
   build. Sys spec headers will be located in
   ``build/projects/libc/include/sys``.

> New Headergen is turned on by default, but if you wanted to use old headergen,

Reviewer: ```suggestion
New Headergen is turned on by default, but if you want to use old headergen,
```

you can include this statement when building: ``-DLIBC_USE_NEW_HEADER_GEN=OFF``

To add a function to the yaml files, you can either manually enter it in the

> yaml file depending on the header spec or enter it through the command line.

Reviewer: ```suggestion
yaml file corresponding to the header it belongs to or add it through the command line.
```

> To enter through the command line:

Reviewer: change to match above

1. Make sure you are in the llvm-project directory.

2. Enter in the command line:
   ``python3 libc/newhdrgen/yaml_to_classes.py
   libc/newhdrgen/yaml/[yaml_file.yaml] --add_function “<return_type>”
   <function_name> “<function_arg1, function_arg2>” <standard> <guard>
   <attribute>``

   Example:
   ``python3 libc/newhdrgen/yaml_to_classes.py
   libc/newhdrgen/yaml/ctype.yaml --add_function “char” example_funtion “int,
   void, const void” stdc example_float example_attribute``
   
   Keep in mind only the return_type and arguments have quotes around them. If
   you do not have any guards or attributes you may enter “null” for both.

3. Check the yaml file that the added function is present. You will also get a
   generated header file with the new addition in the newhdrgen directory to
   examine.

Testing

> -----------------------------

Reviewer: this line should also math the one above it and same for a couple below

New Headergen has an integration test that you may run once you have configured
your CMake within the build directory. In the command line, enter the following:

> “ninja check -newhdrgen”. The integration test is one test that ensures the

Reviewer: here (and elsewhere) make sure that the quote marks you use are `""` instead of `“”`. Also `ninja check-newhdrgen` shouldn't have a space in the middle.

process of yaml to classes to generate headers works properly. If there are any
new additions on formatting headers, make sure the test is updated with the
specific addition.

Integration Test can be found in: ``libc/newhdrgen/tests/test_integration.py``

File to modify if adding something to formatting:
``libc/newhdrgen/tests/expected_output/test_header.h``

Common Errors
-----------------------------
1. ``"/llvm-project/libc/newhdrgen/yaml_to_classes.py", line 67, in yaml_to_classes function_data["return_type"]``

>     If you receive this error or any error pertaining to

Reviewer: The formatting seems a bit odd to me. This might work better if the errors are subheadings and the paragraphs are just normal.

easier cross compatibility.

There are 3 main components of the new Headergen. The first component are the
yaml files that contain all the function header information and are separated by
header specification and standard. The second component are the classes that are
created for each component of the function header: macros, enumerations, types,
function, arguments, and objects. The third component is the python script that
uses the class representation to deserialize yaml files into its specific
components and then reserializes the components into the function header. The
python script also combines the generated header content with header definitions
and extra macro and type inclusions from the .h.def file.

Instructions
------------

Required Versions:

- Python Version: 3.11.8
- CMake Version: 3.20.0

1. Make sure to have `LLVM <https://llvm.org/docs/GettingStarted.html>`_ on your
   system.
2. Make sure you have created a build directory within your LLVM directory.
3. When generating the necessary files by your build make sure to include:
   ``-DLLVM_LIBC_FULL_BUILD=ON`` within the CMake command since new headergen

>    only runs on full-build.

Reviewer 2: I think this is too verbose, except maybe the Python version if you're sure that newhdrgen won't work with Python 3.10 or lower. Could just say that headers are only generated in fullbuild mode otherwise.

4. Once the build is complete, enter in the command line within the build
   directory ``ninja check-newhdrgen`` to ensure that the integration tests are
   passing.
5. Then enter in the command line ``ninja libc`` to generate headers. Headers

>    will be in ``build/projects/libc/include`` or ``build/libc/include`` in a runtime

Reviewer 2: Nit: "runtime" in line 42 exceeds 80 columns, which seems to be the chosen column limit.

   build. Sys spec headers will be located in
   ``build/projects/libc/include/sys``.

New Headergen is turned on by default, but if you want to use old headergen,
you can include this statement when building: ``-DLIBC_USE_NEW_HEADER_GEN=OFF``

To add a function to the yaml files, you can either manually enter it in the
yaml file corresponding to the header it belongs to or add it through the
command line.

To add through the command line:

1. Make sure you are in the llvm-project directory.

2. Enter in the command line:

>    :raw-html:`<br />` ``python3 libc/newhdrgen/yaml_to_classes.py

Reviewer 2: Nit: why ``:raw-html:`<br />` `` followed by an inline code block? Why not a non-inline code block?

specific addition.

Integration Test can be found in: ``libc/newhdrgen/tests/test_integration.py``

File to modify if adding something to formatting:
``libc/newhdrgen/tests/expected_output/test_header.h``

Common Errors
-------------
1. Missing function specific component
   
  Example:
  :raw-html:`<br />` ``"/llvm-project/libc/newhdrgen/yaml_to_classes.py", line
  67, in yaml_to_classes function_data["return_type"]``

  If you receive this error or any error pertaining to
  ``function_data[function_specific_component]`` while building the headers
  that means the function specific component is missing within the yaml files.
  Through the call stack, you will be able to find the header file which has the
  issue. Ensure there is no missing function specific component for that yaml
  header file.

2. CMake Error: require argument to be specified

  Example:
  :raw-html:`<br />` ``CMake Error at:
  /llvm-project/libc/cmake/modules/LLVMLibCHeaderRules.cmake:86 (message):``
  :raw-html:`<br />` ``'add_gen_hdr2' rule requires GEN_HDR to be specified.``
  :raw-html:`<br />` ``Call Stack (most recent call first):
  /llvm-project/libc/include/CMakeLists.txt:22 (add_gen_header2)

>   /llvm-project/libc/include/CMakeLists.txt:62 (add_header_macro)``

Reviewer 2: Nit: why not a single non-inline code block?
